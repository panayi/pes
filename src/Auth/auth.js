import R from 'ramda';
import { createSelector } from 'reselect';
import { FIREBASE_PATH } from '../Firebase/firebase';

// ------------------------------------
// Constants
// ------------------------------------
const FIREBASE_AUTH_PATH = [...FIREBASE_PATH, 'auth'];

// ------------------------------------
// Selectors
// ------------------------------------

// firebaseSelector :: State -> Object
const firebaseAuthSelector = R.pathOr({}, FIREBASE_AUTH_PATH);

// isLoadedSelector :: State -> Maybe(Boolean)
const isLoadedSelector = createSelector(
  firebaseAuthSelector,
  R.compose(
    R.prop('isLoaded'),
    R.defaultTo({}),
  ),
);

// isInitializingSelector :: State -> Maybe(Boolean)
const isInitializingSelector = R.path([...FIREBASE_PATH, 'isInitializing']);

export const uidSelector = createSelector(
  firebaseAuthSelector,
  R.prop('uid'),
);

// hasUidSelector :: State -> Maybe(String)
const hasUidSelector = createSelector(
  uidSelector,
  R.compose(
    R.not,
    R.isNil,
  ),
);

// isAuthenticatingSelector :: State -> Boolean
export const isAuthenticatingSelector = createSelector(
  isInitializingSelector,
  isLoadedSelector,
  R.useWith(R.or, [
    R.identity,
    R.not,
  ]),
);

// isAuthenticatedSelector :: State -> Boolean
// Equals to: !isAuthenticatingSelector && hasUidSelector
export const isAuthenticatedSelector = createSelector(
  isAuthenticatingSelector,
  hasUidSelector,
  R.useWith(R.and, [
    R.not,
    R.identity,
  ]),
);

// isNotAuthenticatedSelector :: State -> Boolean
// Equals to: !isAuthenticatingSelector && !hasUidSelector
export const isNotAuthenticatedSelector = createSelector(
  isAuthenticatingSelector,
  hasUidSelector,
  R.useWith(R.and, [
    R.not,
    R.not,
  ]),
);
