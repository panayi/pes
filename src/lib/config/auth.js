export default {
  google: {
    type: 'popup',
    scopes: 'profile email',
  },
  facebook: {
    type: 'popup',
    scopes: 'public_profile,email,user_friends',
  },
};
