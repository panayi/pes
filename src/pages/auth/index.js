import React from 'react';
import { Route, Switch } from 'react-router-dom';
import needsVisitor from '../../auth/visibility/needsVisitor';
import Layout from '../../layout';
import Login from './login';

const Auth = () => (
  <Layout>
    <Switch>
      <Route exact path="/auth/login" component={Login} />
    </Switch>
  </Layout>
);

export default needsVisitor()(Auth);
