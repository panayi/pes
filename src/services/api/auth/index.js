export const login = credentials => (dispatch, getState, getFirebase) =>
  getFirebase().login(credentials);

export const getProviders = email => (dispatch, getState, getFirebase) =>
  getFirebase()
    .auth()
    .fetchProvidersForEmail(email);
