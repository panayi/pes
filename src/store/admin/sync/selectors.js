import * as R from 'ramda';
import { createSelector } from 'reselect';
import * as constants from './constants';

const SYNC_PATH = ['admin', 'sync'];

const statusSelector = R.path([...SYNC_PATH, 'status']);

const isStatusEqualTo = value => createSelector(
  statusSelector,
  R.equals(value),
);

export const isStatusIdleSelector = isStatusEqualTo(constants.STATUS_IDLE);

export const isStatusSucceededSelector = isStatusEqualTo(constants.STATUS_SUCCEEDED);

const syncedImageSelector = R.path([...SYNC_PATH, 'images']);

const syncedPostsSelector = R.path([...SYNC_PATH, 'posts']);

export const postsSelector = createSelector(
  syncedPostsSelector,
  syncedImageSelector,
  R.useWith(R.merge, [
    R.identity,
    R.map(count => ({ syncedImagesCount: count })),
  ]),
);
