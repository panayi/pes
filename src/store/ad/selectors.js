/* @flow */
import * as R from 'ramda';
import { createSelector } from 'reselect';
import propsSelector from 'utils/propsSelector';
import { PENDING_ADS } from 'services/connectData/types';
import { uidSelector } from '../auth/selectors';

// adImagesPathSelector :: (_, Props) -> Object | Nil
export const adImagesPathSelector = createSelector(
  R.compose(R.prop('adId'), propsSelector),
  adId => `ads/${adId}/images`,
);

// pendingAdImagesSelector :: State -> Object | Nil
export const pendingAdImagesPathSelector = createSelector(
  uidSelector,
  R.unless(R.isNil, uid => `${PENDING_ADS}/${uid}/images`),
);

// pendingAdSelector :: (_, Props) -> Ad | Nil
//   Ad = Object
export const pendingAdSelector = createSelector(
  propsSelector,
  R.prop('pendingAd'),
);
