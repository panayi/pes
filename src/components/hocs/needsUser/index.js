import * as R from 'ramda';
import { connectedReduxRedirect } from 'redux-auth-wrapper/history4/redirect';
import { routerActions } from 'react-router-redux';
import { isAuthenticatedSelector, isAuthenticatingSelector } from 'store/auth/selectors';

export default options => connectedReduxRedirect({
  redirectPath: '/auth/login',
  authenticatedSelector: isAuthenticatedSelector,
  authenticatingSelector: isAuthenticatingSelector,
  redirectAction: routerActions.replace,
  ...R.defaultTo({}, options),
});
