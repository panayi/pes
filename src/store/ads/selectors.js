/* @flow */
import * as R from 'ramda';
import { createSelector } from 'reselect';
import propsSelector from 'utils/propsSelector';
import { PENDING_ADS } from 'services/connectData/types';
import { uidSelector, profilePropSelector } from 'store/auth/selectors';
import * as constants from './constants';

export const createAdPath = ['ad', 'create'];

export const createAdSelector = R.path(createAdPath);

export const isCreateAdIdleSelector = createSelector(
  createAdSelector,
  R.equals(constants.AD_CREATE_IDLE_STATE),
);

export const isCreateAdPendingSelector = createSelector(
  createAdSelector,
  R.equals(constants.AD_CREATE_PENDING_STATE),
);

export const isCreateAdCompletedSelector = createSelector(
  createAdSelector,
  R.equals(constants.AD_CREATE_COMPLETED_STATE),
);

// adImagesPathSelector :: Props -> Object | Nil
export const adImagesPathSelector = createSelector(
  R.compose(R.prop('adId'), propsSelector),
  adId => `ads/${adId}/images`,
);

// pendingAdImagesSelector :: State -> Object | Nil
export const pendingAdImagesPathSelector = createSelector(
  uidSelector,
  R.unless(R.isNil, uid => `${PENDING_ADS}/${uid}/images`),
);

// pendingAdSelector :: Props -> Ad | Nil
//   Ad = Object
export const pendingAdSelector = createSelector(
  propsSelector,
  R.prop('pendingAd'),
);

export const currentUserAdIdsSelector = createSelector(
  profilePropSelector(['ads']),
  R.keys,
);
