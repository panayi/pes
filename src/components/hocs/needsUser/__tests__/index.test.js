import React from 'react';
import { mount } from 'enzyme';
import * as authSelectors from 'store/auth/selectors';
import needsUser from '../index';

describe('[HOC] needsUser', () => {
  authSelectors.isAuthenticatedSelector = jest.fn();
  authSelectors.isAuthenticatingSelector = jest.fn();

  const location = {
    pathname: '/some-user-protected-url',
    search: '',
    hash: '',
  };

  const MyComponent = () => <div>i am a user</div>;
  const Hoc = needsUser({
    redirectPath: '/you-cant-see-user',
  })(MyComponent);
  const { component, store } = withMockStore(<Hoc location={location} />);

  it('should render wrappedComponent correctly', () => {
    authSelectors.isAuthenticatedSelector.mockReturnValueOnce(true);
    authSelectors.isAuthenticatingSelector.mockReturnValueOnce(false);
    const wrapper = mount(component);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render wrappedComponent when isAuthenticatedSelector = true', () => {
    authSelectors.isAuthenticatedSelector.mockReturnValueOnce(true);
    authSelectors.isAuthenticatingSelector.mockReturnValueOnce(false);
    const wrapper = mount(component);
    expect(wrapper.find(MyComponent).exists()).toBe(true);
  });

  it('should not render wrappedComponent and redirect, when isAuthenticatedSelector = false', () => {
    authSelectors.isAuthenticatedSelector.mockReturnValueOnce(false);
    authSelectors.isAuthenticatingSelector.mockReturnValueOnce(false);
    const wrapper = mount(component);
    expect(wrapper.find(MyComponent).exists()).toBe(false);
    expect(store.getActions()).toMatchSnapshot();
  });
});
