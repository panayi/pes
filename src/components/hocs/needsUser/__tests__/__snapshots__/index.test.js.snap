// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[HOC] needsUser should not render wrappedComponent and redirect, when isAuthenticatedSelector = false 1`] = `
Array [
  Object {
    "payload": Object {
      "args": Array [
        Object {
          "hash": null,
          "pathname": "/you-cant-see-user",
          "search": "redirect=%2Fsome-user-protected-url",
        },
      ],
      "method": "replace",
    },
    "type": "@@router/CALL_HISTORY_METHOD",
  },
]
`;

exports[`[HOC] needsUser should render wrappedComponent correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "firebase": Object {
        "INTERNAL": Object {
          "ErrorFactory": [Function],
          "Promise": [Function],
          "createFirebaseNamespace": [Function],
          "createSubscribe": [Function],
          "deepExtend": [Function],
          "extendNamespace": [Function],
          "factories": Object {
            "auth": [Function],
            "database": [Function],
          },
          "node": Object {
            "XMLHttpRequest": [Function],
            "localStorage": Storage {},
            "sessionStorage": Storage {},
          },
          "registerService": [Function],
          "removeApp": [Function],
          "useAsService": [Function],
        },
        "Promise": [Function],
        "SDK_VERSION": "4.5.0",
        "User": [Function],
        "_": Object {
          "authUid": null,
          "config": Object {
            "apiKey": "asdf",
            "authDomain": "asdf",
            "databaseURL": "ws://127.0.1:5000",
            "enableRedirectHandling": false,
            "messagingSenderId": "asdf",
            "storageBucket": "asdf",
            "userProfile": "users",
          },
          "watchers": Object {},
        },
        "__esModule": true,
        "app": [Function],
        "apps": Array [],
        "auth": [Function],
        "database": [Function],
        "default": [Circular],
        "initializeApp": [Function],
      },
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(AuthWrapper(MyComponent))
    location={
      Object {
        "hash": "",
        "pathname": "/some-user-protected-url",
        "search": "",
      }
    }
  >
    <AuthWrapper(MyComponent)
      dispatch={[Function]}
      isAuthenticated={true}
      isAuthenticating={false}
      location={
        Object {
          "hash": "",
          "pathname": "/some-user-protected-url",
          "search": "",
        }
      }
      redirectPath="/you-cant-see-user"
    >
      <MyComponent
        dispatch={[Function]}
        isAuthenticated={true}
        isAuthenticating={false}
        location={
          Object {
            "hash": "",
            "pathname": "/some-user-protected-url",
            "search": "",
          }
        }
        redirectPath="/you-cant-see-user"
      >
        <div>
          i am a user
        </div>
      </MyComponent>
    </AuthWrapper(MyComponent)>
  </Connect(AuthWrapper(MyComponent))>
</Provider>
`;
